src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/WhyCommand/WhyCommandRunner.cs

```diff
@@ -0,0 +1,548 @@
+using System.Threading.Tasks;
+using Microsoft.Build.Evaluation;
+using NuGet.ProjectModel;
+using NuGet.Shared;
+
+namespace NuGet.CommandLine.XPlat
+{
+    internal class WhyCommandRunner : IWhyCommandRunner
+    {
+        private const string ProjectName = "MSBuildProjectName";
+        private const string ProjectAssetsFile = "ProjectAssetsFile";
+
+        // Dependency graph console output symbols
+        private const string ChildNodeSymbol = "├─── ";
+        private const string LastChildNodeSymbol = "└─── ";
+
+        private const string ChildPrefixSymbol = "│    ";
+        private const string LastChildPrefixSymbol = "     ";
+
+        private const string BoldTextEscapeChars = "\u001b[1m";
+        private const string ResetTextEscapeChars = "\u001b[0m";
+
+        /// <summary>
+        /// Executes the 'why' command.
+        /// </summary>
+        /// <param name="whyCommandArgs">CLI arguments for the 'why' command.</param>
+        /// <returns></returns>
+        public Task ExecuteCommand(WhyCommandArgs whyCommandArgs)
+        {
+            var msBuild = new MSBuildAPIUtility(whyCommandArgs.Logger);
+
+            string targetPackage = whyCommandArgs.Package;
+
+            IEnumerable<string> projectPaths = Path.GetExtension(whyCommandArgs.Path).Equals(".sln")
+                                                    ? MSBuildAPIUtility.GetProjectsFromSolution(whyCommandArgs.Path).Where(f => File.Exists(f))
+                                                    : new List<string>() { whyCommandArgs.Path };
+
+            foreach (var projectPath in projectPaths)
+            {
+                Project project = MSBuildAPIUtility.GetProject(projectPath);
+
+                // if the current project is not a PackageReference project, print an error message and continue to the next project
+                if (!MSBuildAPIUtility.IsPackageReferenceProject(project))
+                {
+                    Console.Error.WriteLine(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.Error_NotPRProject,
+                            projectPath));
+
+                    ProjectCollection.GlobalProjectCollection.UnloadProject(project);
+                    continue;
+                }
+
+                string projectName = project.GetPropertyValue(ProjectName);
+                string assetsPath = project.GetPropertyValue(ProjectAssetsFile);
+
+                // if the assets file was not found, print an error message and continue to the next project
+                if (!File.Exists(assetsPath))
+                {
+                    Console.Error.WriteLine(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.Error_AssetsFileNotFound,
+                            projectPath));
+
+                    ProjectCollection.GlobalProjectCollection.UnloadProject(project);
```

Comment:
In general, if you're gonna have something that needs done at the end of a loop, I think it's better to just have it in one spot. 

I think it leads to fewer issues down the line if you don't need to copy around cleanup/recovery code.

Reply:
I've refactored this method, and we only unload at a single point in the code now.